pkg/utils/tsoutil/tso_dispatcher_test.go

```diff
@@ -0,0 +1,201 @@
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+
+	"github.com/tikv/pd/pkg/utils/testutil"
+)
+
+// mockStream is a mock stream used to test TSO request processing
+type mockStream struct {
+	sync.Mutex
+	// delay is used to simulate a long-running processRequest call
+	delay     time.Duration
+	succCount int // Counter for successfully processed requests
+}
+
+// process simulates the process of handling TSO requests
+func (m *mockStream) process(clusterID uint64, _ uint32, _ uint32, _ uint32) (tsoResp, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	m.succCount++
+	// Simulate an error after every 10 successful requests
+	if m.succCount == 10 {
+		// Simulate stream error
+		err := errors.New("simulated stream error")
+		m.succCount = 0
+		return nil, err
+	}
+
+	// If delay is set, sleep for a while to simulate processing time
+	if m.delay > 0 {
+		time.Sleep(m.delay)
+	}
+
+	// Simulate a successful response
+	return &pdpb.TsoResponse{
+		Header: &pdpb.ResponseHeader{
+			ClusterId: clusterID,
+		},
+		Timestamp: &pdpb.Timestamp{
+			Physical: time.Now().UnixMilli(), // Use current time as physical timestamp
+			Logical:  1,                      // Logical timestamp is fixed at 1
+		},
+	}, nil
+}
+
+type mockProtoFactory struct {
+	stream *mockStream
+}
+
+func (m *mockProtoFactory) createForwardStream(_ context.Context, _ *grpc.ClientConn) (stream, context.CancelFunc, error) {
+	if m.stream == nil {
+		return nil, nil, errors.New("no stream available")
+	}
+	return m.stream, func() {}, nil
+}
+
+type mockRequest struct {
+	forwardedHost string
+	clientConn    *grpc.ClientConn
+	count         uint32
+	doneCh        chan struct{}
+	err           error
+}
+
+func (m *mockRequest) getForwardedHost() string {
+	return m.forwardedHost
+}
+
+func (m *mockRequest) getClientConn() *grpc.ClientConn {
+	return m.clientConn
+}
+
+func (m *mockRequest) getCount() uint32 {
+	return m.count
+}
+
+func (m *mockRequest) process(forwardStream stream, count uint32) (tsoResp, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	return forwardStream.process(0, count, 0, 0)
+}
+
+func (m *mockRequest) postProcess(countSum int64, _ int64, _ int64) (int64, error) {
+	close(m.doneCh)
+	return countSum, m.err
+}
+
+type tsoDispatcherTestSuite struct {
+	suite.Suite
+	dispatcher *TSODispatcher
+}
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m, testutil.LeakOptions...)
+}
+
+func TestTSODispatcherTestSuite(t *testing.T) {
+	suite.Run(t, new(tsoDispatcherTestSuite))
+}
+
+func (suite *tsoDispatcherTestSuite) SetupTest() {
+	tsoProxyHandleDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name: "test_tso_proxy_handle_duration_seconds",
+		Help: "Histogram of TSO proxy handle duration",
+	})
+	tsoProxyBatchSize := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name: "test_tso_proxy_batch_size",
+		Help: "Histogram of TSO proxy batch size",
+	})
+	suite.dispatcher = NewTSODispatcher(tsoProxyHandleDuration, tsoProxyBatchSize)
```

Comment:
What if we pass `nil` for `tsoProxyBatchSize`? In that case, we can simply add an `if` check inside `processRequests` instead of mocking unused histograms.